### ISSUE TITLE
@GraphQLNonNull ignored on Input objects when @JsonProperty is used

### ISSUE THREAD

**Issue by EnilPajic** [leangen/graphql-spqr#244]
Hello,

recently the frontend team informed me that on some of our inputs there are no required fields. I've traced down the bug to Jackson's `@JsonProperty` which was added some time ago on constructor parameters (constructor itself was annotated with `@JsonCreator`).

Here is the MVCE that produces no `!` in generated GQL schema (i.e. every field is nullable):
```java
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import graphql.schema.idl.SchemaPrinter;
import io.leangen.graphql.GraphQLSchemaGenerator;
import io.leangen.graphql.annotations.GraphQLArgument;
import io.leangen.graphql.annotations.GraphQLNonNull;
import io.leangen.graphql.annotations.GraphQLQuery;

class Test {
    public static void main(String[] args) {
        GraphQLSchemaGenerator schemaGenerator = new GraphQLSchemaGenerator()
                .withOperationsFromSingleton(new Test());
        System.out.println(new SchemaPrinter().print(schemaGenerator.generate()));
    }
    @GraphQLQuery(name = "RegisterUser")
    public String registerUser(@GraphQLArgument(name = "input") @GraphQLNonNull UserInput input) {
        return input.getFirstName();
    }
}

class UserInput {
    private String firstName;
    private String lastName;

    @JsonCreator
    public UserInput(@JsonProperty("firstName") String firstName,
                     @JsonProperty("lastName") String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    @GraphQLNonNull
    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }
}
```

Schema produced:
```graphql
#Query root
type Query {
  RegisterUser(input: UserInputInput!): String @_mappedOperation(operation : "__internal__")
}

input UserInputInput @_mappedType(type : "__internal__") {
  firstName: String
  lastName: String
}
```

Parameter `input` is okay, but `firstName` in `UserInputInput` is nullable (not required), and there _is_ `@GraphQLNonNull` annotation on getter `getFirstName`.

I've tried following without success:
* adding setters, and annotation it `@GraphQLNonNull UserInput setFirstName (String name)`
* adding setter, but annotating parameter `UserInput setFirstName (@GraphQLNonNull String name)`

Correct schema is produced when `@JsonProperty` is removed, e.g. with this constructor:
```java
    public UserInput(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }
```
(It seems that `@JsonCreator` annotation does not matter)
```graphql
input UserInputInput @_mappedType(type : "__internal__") {
  firstName: String!
  lastName: String
}
```

So, I thought that `@JsonProperty` is considered as "truth source" because it has `JsonProperty.required` field which is by default `false`. But, adding `@JsonProperty(value = "firstName", required = true)` on constructor parameter (or on private field itself, or both) does not change anything.

I'm using `graphql-spqr:0.9.9` and `graphql-java:11.0`.

Is this a bug or is there any workaround for this? Maybe using `@GraphQLInputField`? (I haven't used that before). 

P.S. Another question, I saw on gitter that someone asked how to get rid of `@_mappedOperation` and `@_mappedType`, but answer was vague. Is it possible now, did you find out a way?

Thank you in advance for answer and for this very helpful library!

**Comment by Kirintale** [comment]
I'm following this thread as I'm seeing the same issue. I'm currently implementing GraphQL in an established code base

I have the following subclass
``` 
@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type") 
@JsonSubTypes({ @JsonSubTypes.Type(value = ConcreteNode.class, name = "ConcreteNode") })
@GraphQLInterface(name = "AbstractNode", implementationAutoDiscovery = true)
public abstract class AbstractNode {
    private String name;
    private String id;
    
    /**
     * @return the id
     */
    @GraphQLQuery(name = "id")
    @GraphQLNonNull
    public String getId() {
        return id;
    }

    /**
     * @param id
     *            the id to set
     */
    public void setId(final String id) {
        this.id = id;
    }
    
    // Constructors and other getters and setters
    
}
```
with a concrete implementation 
```
public class ConcreteNode extends AbstractNode {

    private String description;
    
    /**
     * @return the description
     */
    @GraphQLNonNull
    public String getDescription() {
        return description;
    }

    /**
     * @param description
     *            the description to set
     */
    public void setDescription(final String description) {
        this.description = description;
    }
}
```
And the following call
```
@GraphQLMutation(name = "addNode")
    public AbstractNode addNode(@GraphQLArgument(name = "node") @GraphQLNonNull final ConcreteNode node)
            throws Exception {
        // Do stuff
        return node;
    }
```

Annoyingly to get it to run I had to include a variable called `type` on AbstractNode (not shown in example), then when making my graphQL called, I had to state the correct type associated with the class.

```
mutation {
  addNode(node: {name: "nodeName", id: "nodeId", type: "ConcreteNode "}) {
    name
    id
    type
  }
}
```

With that fixed, we have now found that like the original poster, we can't make any of the variables in ConcreteNode mandatory. (It just seems to forget declaration).

Thank you for any response and please keep up the good work.

*PS* If there is a way to get around having to post "type" it would be appreciated.

**Comment by Kirintale** [comment]
Hey OP

I think I've managed to get it working on my end.
What I've had to do is put `@GraphQLNonNull` on the setter.

```
public void setId(@GraphQLNonNull final String id) {
        this.id = id;
    }
```

I believe you may be able to put in on constructors as well, (I could be lying, a lot I'm working out by trial and error). Short of that you might be able to make private setters (if you don't want public ones) with the tag on it.

**Comment by EnilPajic** [comment]
Hai @Kirintale! Thank you for suggesting a workaround!

But, as I've noted before (in my original post), I've tried with annotating setter with `@GraphQLNonNull`. I've  also tried adding the `final` keyword to setter parameter (as in your example), but with same result. Also, adding `@JsonProperty` with `required = true` does not help.
```java
    public UserInput setFirstName(@GraphQLNonNull @JsonProperty(value = "firstName", required = true) final String firstName) {
        this.firstName = firstName;
        return this;
    }
```
Schema produced:

```graphql
#Query root
type Query {
  RegisterUser(input: UserInputInput!): String @_mappedOperation(operation : "__internal__")
}

input UserInputInput @_mappedType(type : "__internal__") {
  firstName: String #missing !
  lastName: String
}
```
Lastly I changed setter from fluent one (returning `UserInput`) to classic one (returning `void`), but the results are the same. I don't know how putting `@GraphQLNonNull` on you setter parameter resolved the issue, but it doesn't work for me. Maybe because of inheritance or something, but I doubt.

But your suggestion to put `@GraphQLNonNull` on **constructor** parameter resolved the issue! This is indeed a workaround, but it is working! Thank you very much!

_(for completeness & other people reading, I'll write down the full "working" example)_

```java
class UserInput {
    private String firstName;
    private String lastName;

    @JsonCreator
    public UserInput(@GraphQLNonNull @JsonProperty("firstName") String firstName, //added @GraphQLNonNull
                     @JsonProperty("lastName") String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    //@GraphQLNonNull //in this workaround this is not needed
    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }
}
```


I'll leave the issue open because I assume this is still a bug.

**Comment by Kirintale** [comment]
Ace glad putting it on the constructor worked!

Strange that putting it on the setter worked for me but not you, maybe that is related to @JsonCreator.

I suspect the following occurs when I deserialize data.
1. The data is created using a default constructor.
1. The setters and reflection are used to apply set the variables. So putting `@GraphQLNonNull ` on my setters worked.

In contrast, if I am understanding what `@JsonCreator` does correctly. When you run your code
1. The constructor labelled with `@JsonCreator` is used and any `@GraphQLNonNull ` instance is called as well. 
1. Your setters aren't called (thus it wasn't required there).

I'll test it out in my code to confirm.

**Comment by EnilPajic** [comment]
Hi!

Thank you for testing this out. Your understanding of `@JsonCreator` is right, if present, setters won't be called.

But I've noticed something: if `@JsonCreator` is removed, the incorrect schema is produced again, e.g.  with this code:
```java
class UserInput {
    private String firstName;
    private String lastName;

    //no JsonCreator
    public UserInput(@JsonProperty("firstName") String firstName,
                     @JsonProperty("lastName") String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }
    
    @GraphQLNonNull
    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }
}
```

So, in this case, it seems that `@JsonCreator` does not matter (as stated in my original post). But what matters is the `@JsonProperty` declaration on constructor parameters. If I remove them, everything works fine (with `@GraphQLNonNull` on getter):
```java
class UserInput {
    private String firstName;
    private String lastName;


    //no matter if @JsonCreator is present or not, schema is correct
    public UserInput(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    @GraphQLNonNull
    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }
}
```

The above code indeed produces correct schema, so I assume that `@JsonProperty` is the reason?
And if `@GraphQLNonNull` is working without `@JsonProperty` but not with it, this may be a bug?

**Comment by Kirintale** [comment]
Yeah that's interesting. I made another simple class which I was using to test SPQR and didn't have any reference to any @JSon functionality and I think I discovered the same as you. I initially was trying @GraphQLNonNull on the getter, so it took me a while to even think to put it on the setter what I started to have an issue.

Can you try creating a default constructor with a print statement in it and see it gets hit?

There is something interesting going on in the background, just not entirely certain what. I'm generally treating the SPQR code base as a bit of black box when working stuff out. I sometimes look into its actual code, but its a smart and complex bit of kit and I only have so much time.

You're right it may be a bug or at least something worth getting feedback on.

**Comment by EnilPajic** [comment]
>Can you try creating a default constructor with a print statement in it and see it gets hit?

Tried combination of: only two-param constructor, only default constructor and both, with different "printings" inside them, and nothing was printed. That is logical because in my example only GQL schema is produced/printed which is done via reflection, so no deserialization is occurring. (I didn't call the GraphQL queries, no need for that at all because the produced schema is wrong, not the query result).

I agree with you, this is something worth of feedback.

**Comment by kaqqao** [comment]
Hey everyone. Sorry for the silence on my part, I'm away on a business trip these days.

As for the reason the annotation is apparently ignored... When inputs are mapped using Jackson, the logic is as follows:

 * Find the explicitly `@GraphQLInputField` annotated element
 * Find the primary mutator that Jackson will use to deserialize the values
 * Find the field belonging to the property
 * Merge the types of all 3

So for the case such as:
```java
@JsonCreator
public UserInput(@JsonProperty("firstName") String firstName,
                 @JsonProperty("lastName") String lastName) {
    this.firstName = firstName;
    this.lastName = lastName;
}

@GraphQLNonNull
public String getFirstName() {
    return firstName;
}
```

There's no explicitly annotated element, and the constructor is the primary mutator (that's what will be used to deserialize), so only the field type and the ctor parameter type get merged (getter is not involved), thus the annotation is ignored.

To force the logic to take the getter into account, you can mark it with `@GraphQLInputField`. Or move `@GraphQLNonNull` to the ctor param or the field.

The reason it works the way it does is to avoid the situation where a bunch of unrelated language elements have to be reasoned about at the same time to know what the mapping will be. The field is always taken into account to allow the usage of things like Lombok. The primary mutator is taken into account because that is what will actually be used by Jackson. The only remaining element is the one the user explicitly decides to involve.

> P.S. Another question, I saw on gitter that someone asked how to get rid of @_mappedOperation and @_mappedType, but answer was vague. Is it possible now, did you find out a way?

SPQR uses those to provide some its functionality, so you can't get rid of them completely without tricky repercussions... If you meant how to avoid printing them, there were changes in the `SchemaPrinter` that should allow it, but that's coming from graphql-java so I have no control over it.

I'll read thought the rest of the thread as soon as I get a chance and respond to whatever else there is...

**Comment by kaqqao** [comment]
@Kirintale 
The `type` thing is puzzling me. There's a feature in SPQR that automatically deals with abstract inputs by adding a discriminator enum field called `_type`, but you need to explicitly enable it. Plus, your input isn't even  abstract. So I don't think whatever is making you add the `type` field is coming from SPQR itself, but rather some customization. Unless you're using some very old SPQR version, I don't clearly remember how it used to work there... In that, please, please, please upgrade to 0.9.9.

**Comment by kaqqao** [comment]
@Kirintale Oh, it's actually this line here that adds the `type` field:
```java
@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type") 
```
You're adding it explicitly, so isn't the behavior you see expected?

**Comment by kaqqao** [comment]
@EnilPajic I've opened a new issue for `JsonProperty#required`: https://github.com/leangen/graphql-spqr/issues/287

**Comment by Kirintale** [comment]
Thanks for the responses.

TLDR; After some more research and testing I've got no issue with declaring the `type` anymore (I think), its actually advantageous to my code base.

Long response:
(Apologies I'm about to derail the topic talking about type)
In short, yes, its pretty much behaving the way I expect, I just had to learn more about GraphQL to correct my expectation ^_^ 

What happened was:
```
@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type") 
```
Was required to ensure that Jackson deserialised complex objects correct (this is an established code base where we are experimenting changing the API to use GraphQL rather than just rest calls).
What this meant was that when ingesting JSON files `type` was required in any ingested JSON. When I was reading up about GraphQL I stumbled on some posts where people said you can't pass through an abstract class as a parameter to a GraphQL method. So it initially seemed odd that I had to declare `type` when casting directly to the subclass (though I now see this was caused by the @JSonTypeInfo flag). Moreover, it also meant that I needed to included type member variable in the class (to ensure it was added to the schema).

So my class ended up looking like (note the new variable and use of EXISTING_PROPERTY)
```
@JsonTypeInfo(use = Id.NAME, include = As.EXISTING_PROPERTY, property = "type") 
@JsonSubTypes({ @JsonSubTypes.Type(value = ConcreteNode.class, name = "ConcreteNode") })
@GraphQLInterface(name = "AbstractNode", implementationAutoDiscovery = true)
public abstract class AbstractNode {
    private String name;
    private String id;
    private String type = this.getClass().getSimpleName();

// Getters, setters, etc...
}
```

Thankfully I've since found out (as you just mentioned) that SPQR allows you to have abstract classes as parameters and provided I pass through `type` it will correctly cast the inputted query to the correct class. So I'm happy with the code changes I have had to do, as I had previously feared I would have to create a method to accept each  concrete implementation individually. Hope that makes sense.

**Comment by kaqqao** [comment]
SPQR basically lets Jackson handle the deserialization in whatever way it's configured. If you want it to auto-configure Jackson for polymorphic deserialization, you can use `generator.withAbstractInputTypeResolution()`.

@EnilPajic Do you reckon this issue has been sufficiently addressed by my explanation above? Or do you propose some changes?

**Comment by EnilPajic** [comment]
@kaqqao Thank you for explanation, it makes sense. I'll use `@GraphQLInputField`, but it is nice to know that workaround(s) exist (ctor parameter annotation). Also, it'd be nice if JsonProperty#required is respected, as you pointed in issue. I'll close the issue now.

**Comment by Kirintale** [comment]
Thank you @kaqqao 

I need to update my notes on SQPR now ^_^
